# **KLASIFIKASI KELAS JAMUR DAPAT DIMAKAN DAN BERACUN DENGAN METODE RANDOM FOREST**

disini saya akan melakukan klasifikasi pada data mushroom untuk mengetahui kelas jamur beracun(posinous/p) atau dapat dimakan(edible/e) dengan ciri/karakteristik fisik yang dimiliki.

# **Deskripsi Data**

data yang digunakan di ambil dari website UCI REPOSITORY : https://archive.ics.uci.edu/dataset/73/mushroom
dengan 23 kolom (fitur dan Label) dan memiliki 8124 baris data, sesuai dengan deskripsi dataset pada UCI dijelaskan dataset mushroom berupa dataset kategorikal dan memiliki missing value di fitur stalk-root yang mana missing valuenya di isi dengan nilai "?". terkait missing value kita akan lakukan pengisian dengan modus data yang berlabel sama dengan label missing value. dilakukan dengan modus karna data ini bertipe kategorikal.

**FITUR-FITUR PADA DATASET**

0. Label : edible=e, poisonous=p
1. cap-shape: bell=b,conical=c,convex=x,flat=f, knobbed=k,sunken=s
2. cap-surface:fibrous=f,grooves=g,scaly=y,smooth=s
3. cap-color:brown=n,buff=b,cinnamon=c,gray=g,green=r, pink=p,purple=u,red=e,white=w,yellow=y
4. bruises?: bruises=t,no=f
5. odor:almond=a,anise=l,creosote=c,fishy=y,foul=f, musty=m,none=n,pungent=p,spicy=s
6. gill-attachment:attached=a,descending=d,free=f,notched=n
7. gill-spacing: close=c,crowded=w,distant=d
8. gill-size:  broad=b,narrow=n
9. gill-color: black=k,brown=n,buff=b,chocolate=h,gray=g, green=r,orange=o,pink=p,purple=u,red=e, white=w,yellow=y
10. stalk-shape:enlarging=e,tapering=t
11. stalk-root:bulbous=b,club=c,cup=u,equal=e, rhizomorphs=z,rooted=r,missing=?
12. stalk-surface-above-ring: fibrous=f,scaly=y,silky=k,smooth=s
13. stalk-surface-below-ring: fibrous=f,scaly=y,silky=k,smooth=s
14. stalk-color-above-ring:   brown=n,buff=b,cinnamon=c,gray=g,orange=o, pink=p,red=e,white=w,yellow=y
15. stalk-color-below-ring: brown=n,buff=b,cinnamon=c,gray=g,orange=o, pink=p,red=e,white=w,yellow=y
16. veil-type: partial=p,universal=u
17. veil-color:brown=n,orange=o,white=w,yellow=y
18. ring-number:none=n,one=o,two=t
19. ring-type: cobwebby=c,evanescent=e,flaring=f,large=l, none=n,pendant=p,sheathing=s,zone=z
20. spore-print-color: black=k,brown=n,buff=b,chocolate=h,green=r, orange=o,purple=u,white=w,yellow=y
21. population:abundant=a,clustered=c,numerous=n, scattered=s,several=v,solitary=y
22. habitat:grasses=g,leaves=l,meadows=m,paths=p, urban=u,waste=w,woods=d



# **PROSES PEMBUATAN MODEL**

**==> TAHAP DATA PRE-PROCESSING <==**

**Import Library**

disini akan menggunakan beberapa library, namun yang pertama kita memerlukan library pandas untuk menampilkan data dalam dataframe

In [1]:
import pandas as pd

**Menampilkan data**

In [2]:
data=pd.read_csv('https://raw.githubusercontent.com/LuluatulMaknunah21-048/PSD/main/agaricus-lepiota.data',header=None)
data

0  1  2  3  4  5  6  7  8  9   ... 13 14 15 16 17 18 19 20 21 22
0     p  x  s  n  t  p  f  c  n  k  ...  s  w  w  p  w  o  p  k  s  u
1     e  x  s  y  t  a  f  c  b  k  ...  s  w  w  p  w  o  p  n  n  g
2     e  b  s  w  t  l  f  c  b  n  ...  s  w  w  p  w  o  p  n  n  m
3     p  x  y  w  t  p  f  c  n  n  ...  s  w  w  p  w  o  p  k  s  u
4     e  x  s  g  f  n  f  w  b  k  ...  s  w  w  p  w  o  e  n  a  g
...  .. .. .. .. .. .. .. .. .. ..  ... .. .. .. .. .. .. .. .. .. ..
8119  e  k  s  n  f  n  a  c  b  y  ...  s  o  o  p  o  o  p  b  c  l
8120  e  x  s  n  f  n  a  c  b  y  ...  s  o  o  p  n  o  p  b  v  l
8121  e  f  s  n  f  n  a  c  b  n  ...  s  o  o  p  o  o  p  b  c  l
8122  p  k  y  n  f  y  f  c  n  b  ...  k  w  w  p  w  o  e  w  v  l
8123  e  x  s  n  f  n  a  c  b  y  ...  s  o  o  p  o  o  p  o  c  l

[8124 rows x 23 columns]

data yang diambil dari uci bukan data csv maka dari itu saya transformasi menjadi data csv supaya menjadi data tabular.

**Menyimpan data kedalam format CSV**

In [3]:
data.to_csv('jamur.csv',index=False)

**Menampilkan data yang sudah di simpan dalam bentuk CSV**

In [4]:
jamur=pd.read_csv('https://raw.githubusercontent.com/LuluatulMaknunah21-048/PSD/main/jamur.csv')
jamur

0  1  2  3  4  5  6  7  8  9  ... 13 14 15 16 17 18 19 20 21 22
0     p  x  s  n  t  p  f  c  n  k  ...  s  w  w  p  w  o  p  k  s  u
1     e  x  s  y  t  a  f  c  b  k  ...  s  w  w  p  w  o  p  n  n  g
2     e  b  s  w  t  l  f  c  b  n  ...  s  w  w  p  w  o  p  n  n  m
3     p  x  y  w  t  p  f  c  n  n  ...  s  w  w  p  w  o  p  k  s  u
4     e  x  s  g  f  n  f  w  b  k  ...  s  w  w  p  w  o  e  n  a  g
...  .. .. .. .. .. .. .. .. .. ..  ... .. .. .. .. .. .. .. .. .. ..
8119  e  k  s  n  f  n  a  c  b  y  ...  s  o  o  p  o  o  p  b  c  l
8120  e  x  s  n  f  n  a  c  b  y  ...  s  o  o  p  n  o  p  b  v  l
8121  e  f  s  n  f  n  a  c  b  n  ...  s  o  o  p  o  o  p  b  c  l
8122  p  k  y  n  f  y  f  c  n  b  ...  k  w  w  p  w  o  e  w  v  l
8123  e  x  s  n  f  n  a  c  b  y  ...  s  o  o  p  o  o  p  o  c  l

[8124 rows x 23 columns]

**Menampilkan nama kolom/fitur**

In [5]:
column_names =jamur.columns
column_names

Index(['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12',
       '13', '14', '15', '16', '17', '18', '19', '20', '21', '22'],
      dtype='object')

karna data yang saya ambil tidak langsung dalam format csv atau bertype tabular dan ketika saya tampilkan datanya nama kolom tidak langsung terbaca maka saya akan merubah nama kolom dari angka menjadi nama kolom yang sesuai dengan deskripsi data pada website UCI repository.

**Mengganti nama kolom menjadi nama kolom sesuai dengan di UCI Repository**

In [6]:
jamur.rename(columns={'0': 'Label', '1': 'cap-shape','2':'cap-surface','3':'cap-color','4':'bruises','5':'odor',
                     '6': 'gill-attachment', '7': 'gill-spacing','8':'gill-size','9':'gill-color','10':'stalk-shape',
                     '11': 'stalk-root', '12': 'stalk-surface-above-ring','13':'stalk-surface-below-ring','14':'stalk-color-above-ring','15':'stalk-color-below-ring',
                     '16': 'veil-type', '17': 'veil-color','18':'ring-number','19':'ring-type','20':'spore-print-color',
                     '21': 'population', '22': 'habitat'},inplace=True)

**Menampilkan Kolom yang sudah dirubah**

In [7]:
jamur

Label cap-shape cap-surface cap-color bruises odor gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
...    ...       ...         ...       ...     ...  ...             ...   
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
0              o         p                 k          s       u  
1              o         p                 n          n       g  
2              o         p                 n          n       m  
3              o         p                 k          s       u  
4              o         e                 n          a       g  
...          ...       ...               ...        ...     ...  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[8124 rows x 23 columns]

**Menyimpan data dalam format CSV (data yang telah diganti nama kolom)**

In [8]:
jamur.to_csv('datajamur.csv',index=False)

**Menampilkan data yang telah disimpan sebelumnya**

In [9]:
df=pd.read_csv('https://raw.githubusercontent.com/LuluatulMaknunah21-048/PSD/main/datajamur.csv')
df

Label cap-shape cap-surface cap-color bruises odor gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
...    ...       ...         ...       ...     ...  ...             ...   
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
0              o         p                 k          s       u  
1              o         p                 n          n       g  
2              o         p                 n          n       m  
3              o         p                 k          s       u  
4              o         e                 n          a       g  
...          ...       ...               ...        ...     ...  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[8124 rows x 23 columns]

**Mengisi Missing Value**

bagian ini kita sudah memasuki tahap pengisian missing value yang ada pada kolom stalk-root, dan mengisi dengan nilai modus dari setiap label sesuai dengan label yang dimiliki oleh missing value

**Melihat Jumlah Missing Value sesuai dengan Label**




In [10]:
summary_data = df.groupby(['Label', 'stalk-root']).size().reset_index(name='jumlah_data')
summary_data

Label stalk-root  jumlah_data
0     e          ?          720
1     e          b         1920
2     e          c          512
3     e          e          864
4     e          r          192
5     p          ?         1760
6     p          b         1856
7     p          c           44
8     p          e          256

**Melihat Modus dari label e dan p pada kolom stalk-root**

In [11]:
moduse= df[df['Label'] == 'e']['stalk-root'].mode().iloc[0]
modusp= df[df['Label'] == 'p']['stalk-root'].mode().iloc[0]
print("Modus stalk-root di kelas e adalah",moduse)
print("Modus stalk-root di kelas p adalah",modusp)

Modus stalk-root di kelas e adalah b
Modus stalk-root di kelas p adalah b


**Mengganti nilai ? dengan nilai modus**

In [12]:
# Mengganti nilai '?'
df.loc[df['Label'] == 'e', 'stalk-root'] = df.loc[df['Label'] == 'e', 'stalk-root'].replace('?', moduse)
df.loc[df['Label'] == 'p', 'stalk-root'] = df.loc[df['Label'] == 'p', 'stalk-root'].replace('?', modusp)

**Memastikan apakah pengubahan nilai missing value sudah berhasil atau belum**

In [13]:
missing = df.groupby(['Label', 'stalk-root']).size().reset_index(name='jumlah_data')
missing

Label stalk-root  jumlah_data
0     e          b         2640
1     e          c          512
2     e          e          864
3     e          r          192
4     p          b         3616
5     p          c           44
6     p          e          256

**NORMALISASI**

karna data yang kita miliki adalah data kategorikal maka data akan dilakukan normalisasi dengan one-hot encoding. One-hot encoding adalah teknik representasi data kategorikal dalam bentuk vektor biner, di mana hanya satu bit bernilai 1 sedangkan yang lainnya bernilai 0. Ini digunakan ketika kita memiliki variabel kategorikal dan ingin mengonversinya menjadi bentuk yang dapat digunakan oleh model machine learning yang memerlukan input numerik.

Contoh one-hot encoding:
Jika kita memiliki variabel kategorikal "Warna" dengan nilai {Merah, Hijau, Biru}, maka setelah one-hot encoding, setiap nilai akan direpresentasikan sebagai vektor biner:

Merah: [1, 0, 0]
Hijau: [0, 1, 0]
Biru: [0, 0, 1]

**Melakukan drop pada kolom label karna kolom label tidak akan dilakukan normalisasi one-hot encoding**

In [14]:
#tidak akan melakukan one-hot encoding pada data label
label= ['Label']
df_label= df[label]
df = df.drop(columns=label)

**Melakukan normaliasi pada fitur serta menambahkan kembali data label**

In [15]:
# Melakukan One-Hot Encoding pada semua kolom
df_encoded = pd.get_dummies(df)
# Menggabungkan DataFrame hasil One-Hot Encoding dengan label
df_final = pd.concat([df_label, df_encoded], axis=1)
df_final

Label  cap-shape_b  cap-shape_c  cap-shape_f  cap-shape_k  cap-shape_s  \
0        p            0            0            0            0            0   
1        e            0            0            0            0            0   
2        e            1            0            0            0            0   
3        p            0            0            0            0            0   
4        e            0            0            0            0            0   
...    ...          ...          ...          ...          ...          ...   
8119     e            0            0            0            1            0   
8120     e            0            0            0            0            0   
8121     e            0            0            1            0            0   
8122     p            0            0            0            1            0   
8123     e            0            0            0            0            0   

      cap-shape_x  cap-surface_f  cap-surface_g  cap-surface_s  ...  \
0               1              0              0              1  ...   
1               1              0              0              1  ...   
2               0              0              0              1  ...   
3               1              0              0              0  ...   
4               1              0              0              1  ...   
...           ...            ...            ...            ...  ...   
8119            0              0              0              1  ...   
8120            1              0              0              1  ...   
8121            0              0              0              1  ...   
8122            0              0              0              0  ...   
8123            1              0              0              1  ...   

      population_s  population_v  population_y  habitat_d  habitat_g  \
0                1             0             0          0          0   
1                0             0             0          0          1   
2                0             0             0          0          0   
3                1             0             0          0          0   
4                0             0             0          0          1   
...            ...           ...           ...        ...        ...   
8119             0             0             0          0          0   
8120             0             1             0          0          0   
8121             0             0             0          0          0   
8122             0             1             0          0          0   
8123             0             0             0          0          0   

      habitat_l  habitat_m  habitat_p  habitat_u  habitat_w  
0             0          0          0          1          0  
1             0          0          0          0          0  
2             0          1          0          0          0  
3             0          0          0          1          0  
4             0          0          0          0          0  
...         ...        ...        ...        ...        ...  
8119          1          0          0          0          0  
8120          1          0          0          0          0  
8121          1          0          0          0          0  
8122          1          0          0          0          0  
8123          1          0          0          0          0  

[8124 rows x 117 columns]

setelah melakukan normalisasi one-hot encoding kolom yang semula 23 kolom (label, 22 fitur) menjadi 117 kolom ( label, 116 fitur).

**Melihat data kolom terbaru setelah normalisasi**

In [16]:
df_final.columns.tolist()

['Label',
 'cap-shape_b',
 'cap-shape_c',
 'cap-shape_f',
 'cap-shape_k',
 'cap-shape_s',
 'cap-shape_x',
 'cap-surface_f',
 'cap-surface_g',
 'cap-surface_s',
 'cap-surface_y',
 'cap-color_b',
 'cap-color_c',
 'cap-color_e',
 'cap-color_g',
 'cap-color_n',
 'cap-color_p',
 'cap-color_r',
 'cap-color_u',
 'cap-color_w',
 'cap-color_y',
 'bruises_f',
 'bruises_t',
 'odor_a',
 'odor_c',
 'odor_f',
 'odor_l',
 'odor_m',
 'odor_n',
 'odor_p',
 'odor_s',
 'odor_y',
 'gill-attachment_a',
 'gill-attachment_f',
 'gill-spacing_c',
 'gill-spacing_w',
 'gill-size_b',
 'gill-size_n',
 'gill-color_b',
 'gill-color_e',
 'gill-color_g',
 'gill-color_h',
 'gill-color_k',
 'gill-color_n',
 'gill-color_o',
 'gill-color_p',
 'gill-color_r',
 'gill-color_u',
 'gill-color_w',
 'gill-color_y',
 'stalk-shape_e',
 'stalk-shape_t',
 'stalk-root_b',
 'stalk-root_c',
 'stalk-root_e',
 'stalk-root_r',
 'stalk-surface-above-ring_f',
 'stalk-surface-above-ring_k',
 'stalk-surface-above-ring_s',
 'stalk-surface-abov

**Melihat jumlah data berdasrkan label**

In [17]:
jumlah= df_final['Label'].value_counts()
jumlah

e    4208
p    3916
Name: Label, dtype: int64

hal ini untuk memastikan apakah data tidak seimbang atau sudah seimbang

**Menyimpan Data yang sudah di normalisasi**

In [18]:
df_final.to_csv('datanorm.csv',index=False)

**==> TAHAP PELATIHAN DATA UNTUK PEMBUATAN MODEL<==**

**Import library random forest,split data, serta accuracy score**

In [19]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

**Data yang digunakan data yang telah di normalisasi**

In [20]:
data= pd.read_csv('https://raw.githubusercontent.com/LuluatulMaknunah21-048/PSD/main/datanorm.csv')
X = data.drop('Label', axis=1)
y = data['Label']

**SPLITTING DATA**

data di bagi menjadi 80,20 dimana 80% data latih dan 20% data uji dengan menggunakan random state 42

In [21]:
# Membagi data menjadi set pelatihan dan uji
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [22]:
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

**Prediksi dengan algoritma random forest**

In [23]:
y_pred = model.predict(X_test)
y_pred

array(['e', 'p', 'p', ..., 'p', 'p', 'p'], dtype=object)

**MELIHAT AKURASI DARI PELATIHAN**

In [24]:
accuracy = accuracy_score(y_test, y_pred)
accuracy

1.0

karna setelah proses normalisasi menghasilkan 117 fitur maka akan dilakukan PCA untuk mempercepat proses komputasi

# **PCA(PRINCIPAL COMPONENT ANALYSIS)**

**Import library PCA dan pickle**

In [25]:
import pickle
from pickle import dump
from sklearn.decomposition import PCA as sklearnPCA

**Proses PCA**

In [26]:
sklearn_pca = sklearnPCA(n_components=7)
X_train_pca= sklearn_pca.fit_transform(X_train)
type(X_train_pca)

numpy.ndarray

n_components yang digunakan 7, n_componets PCA dilakukan dengan mengurangi dari yang tertinggi bukan dimulai dari yang terendah. misal kamu punya 117 fitur maka n_components dimulai dari 107,97 dst..

**Menyimpan mesin yang digunakan untuk PCA data latih**

In [27]:
dump(sklearn_pca, open('PCA7.pkl', 'wb'))

**Load model PCA data latih untuk digunakan pada data testing**

In [28]:
with open('PCA7.pkl', 'rb') as pca:
   loadpca= pickle.load(pca)

**Menerapkan PCA pada data test**

In [29]:
X_test_pca=loadpca.transform(X_test)
X_test_pca.shape

(1625, 7)

**Import Library random forest**

In [30]:
from sklearn.ensemble import RandomForestClassifier
clfr = RandomForestClassifier(n_estimators=100, random_state=42)
clfr.fit(X_train_pca, y_train)

RandomForestClassifier(random_state=42)

**Prediksi**

In [31]:
y_prediksi = clfr.predict(X_test_pca)
y_prediksi

array(['e', 'p', 'p', ..., 'p', 'p', 'p'], dtype=object)

**Menampilkan Akurasi yang dihasilkan**

In [32]:
acc_pca= accuracy_score(y_test,y_prediksi)
print("Akurasi:",acc_pca)

Akurasi: 1.0


**Menyimpan model final**

In [33]:
pickle.dump (clfr, open ('jamur.pkl', 'wb'))

model yang disimpan dengan nama jamur.pkl akan digunakan untuk implementasi data baru yang akan di input dengan user interface streamlit berikut ini.